### PR TITLE
Improve _sqnorm

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

File: chainer/optimizer.py [hunk 1]
```diff
@@ -7,11 +7,21 @@
...
-    with cuda.get_device(x):
-        x = x.ravel()
-        return float(x.dot(x))
+def _sum_sqnorm(arr):
+    sum = {}
+    for x in arr:
+        with cuda.get_device(x) as dev:
+            x = x.ravel()
+            s = x.dot(x)
+        id = int(dev)
```
Comment (beam2d) [inline chainer/optimizer.py]: It hides the builtin function `id`. Use a different name (e.g. `dev_id`).

File: chainer/optimizer.py [hunk 2]
```diff
@@ -7,11 +7,21 @@
 import chainer.link as link_module
 
 
-def _sqnorm(x):
-    # TODO(delta2323): Make it public function and move it to common directory.
-    with cuda.get_device(x):
-        x = x.ravel()
-        return float(x.dot(x))
+def _sum_sqnorm(arr):
+    sum = {}
```
Comment (beam2d) [inline chainer/optimizer.py]: It hides the builtin function `sum`. Use a different name (e.g. `sq_sum`).

File: chainer/optimizer.py [hunk 3]
```diff
@@ -7,11 +7,21 @@
...
+    for x in arr:
+        with cuda.get_device(x) as dev:
+            x = x.ravel()
+            s = x.dot(x)
+        id = int(dev)
+        if id in sum:
```
Comment (beam2d) [inline chainer/optimizer.py]: How about using defaultdict?

**Comment by beam2d** [comment]
Would you add multi-gpu tests (both with and without cpu)?

**Comment by okuta** [comment]
@beam2d I fixed the code.

**Comment by beam2d** [comment]
LGTM!

**Pull request merged**